profiles/audit_log.py

```diff
@@ -37,56 +86,109 @@ def _profile_part(instance):
+            # If the Profile is about to get deleted,
+            # need to store the User before it gets deleted too.
+            profile_loggables["user"] = profile.user
+        profile_loggables["profile"] = profile
+        profile_loggables["parts"].add((action, _profile_part(instance)))
+
+
+def _produce_json_logs(
+    current_user, current_time, service, client_id, ip_address, audit_loggables
+):
+    logger = logging.getLogger("audit")
+
+    date_time_epoch = int(current_time.timestamp() * 1000)
+    date_time = f"{current_time.replace(tzinfo=None).isoformat(sep='T', timespec='milliseconds')}Z"
+
+    actor_obj = {}
+
+    if service:
+        actor_obj["service_name"] = service.name
+
+    if client_id:
+        actor_obj["client_id"] = client_id
+
+    if ip_address:
+        actor_obj["ip_address"] = ip_address
+
+    _format_user_data(actor_obj, current_user)
+
+    for profile_id, data in audit_loggables.items():
+        target_user = data.get("user")
+
+        actor_obj["role"] = _resolve_role(current_user, target_user)
+
+        target_obj = {
+            "id": str(profile_id),
         }
+        _format_user_data(target_obj, target_user)
+
+        for action, profile_part in data["parts"]:
+            target_obj["type"] = profile_part
+
+            message = {
+                "audit_event": {
+                    "origin": "PROFILE-BE",
+                    "status": "SUCCESS",
+                    "date_time_epoch": date_time_epoch,
+                    "date_time": date_time,
+                    "actor": actor_obj,
+                    "operation": action,
+                    "target": target_obj,
+                }
+            }
 
-        _format_user_data(message["audit_event"], "actor", current_user)
+            logger.info(json.dumps(message))
 
-        _format_user_data(message["audit_event"], "target", target_user)
 
-        service = get_current_service()
-        if service:
-            message["audit_event"]["actor"]["service_name"] = service.name
-        client_id = get_current_client_id()
-        if client_id:
-            message["audit_event"]["actor"]["client_id"] = client_id
+def _commit_audit_logs():
+    request = _get_current_request()
 
-        ip_address = get_original_client_ip()
-        if ip_address:
-            message["audit_event"]["actor"]["ip_address"] = ip_address
+    audit_loggables = getattr(request, "_audit_loggables", None)
+    if not audit_loggables:
+        return
+
+    del request._audit_loggables
+
+    current_time = datetime.now(tz=timezone.utc)
+
+    current_user = _get_current_user()
+    service = _get_current_service()
+    client_id = _get_current_client_id()
+    ip_address = _get_original_client_ip()
+
+    profiles = [log_data["profile"] for log_data in audit_loggables.values()]
+    for user in (
+        User.objects.select_related("profile").filter(profile__in=profiles).all()
+    ):
+        audit_loggables[user.profile.id]["user"] = user
```

Comment:
This could potentially use the `QuerySet.values()` or `.values_list()` methods as there isn't really need for the full objects.

Reply:
Changed to use `QuerySet.values()`.